File: 0648-replace-words/0648-replace-words.java
class Solution {
    public String replaceWords(List<String> dictionary, String sentence) {
        Set<String> dictSet = new HashSet<>(dictionary);
        int index = 0;
        StringBuilder result = new StringBuilder();
        
        while(index < sentence.length()) {
            int rightIndex = index + 1;
            boolean addedString = false;
            
            while(rightIndex < sentence.length() && sentence.charAt(rightIndex) != ' '){
                String substring = sentence.substring(index, rightIndex);
                
                if(dictSet.contains(substring) && !addedString){
                    result.append(" ");
                    result.append(substring);
                    addedString = true;
                }
                
                rightIndex++;
            }
            
            if(!addedString){
                String substring = sentence.substring(index, rightIndex);
                result.append(" ");
                result.append(substring);
            }
            
            while(index < sentence.length() && sentence.charAt(index) != ' '){
                index++;
            }
            
            index++;
        }
        
        return result.toString().substring(1);
    }
}

File: 0648-replace-words/0648-replace-words.js
/**
 * @param {string[]} dictionary
 * @param {string} sentence
 * @return {string}
 */
var replaceWords = function(dictionary, sentence) {
    let result = "";
    let stringArray = sentence.split(" ");
    
    for(let index = 0; index < stringArray.length; index++) {
        let candidate = "";
        let found = false;
        
        for(let subIndex = 0; subIndex < stringArray[index].length; subIndex++) {
            candidate += stringArray[index][subIndex];
            
            if(dictionary.includes(candidate)) {
                result += candidate;
                found = true;
                break;
            }
        }
        
        if(!found) {
            result += stringArray[index];
        }
        
        if(index !== stringArray.length - 1) {
            result += " ";
        }
    }
    
    return result;
};